### ISSUE TITLE
make_dilated method is incorrect

### ISSUE THREAD

**Issue by howard-mahe** [qubvel/segmentation_models.pytorch#183]
Hello @qubvel, thank you for this amazing project. This is frankly the best semantic segmentation library based on PyTorch I've worked with so far.

I want to catch your eye on a issue in `make_dilated` (which call `replace_strides_with_dilation` method). This issue affects `DeepLabV3` and `PAN` which indeed calls `self.encoder.make_dilated`. When we make dilated a block of residual units (e.g. `ResNet.layer4`), the goal is to increase the spatial resolution of this block while preserving the receptive field (in blue in my diagram) of the non-dilated version for each neurons – as long as we initialize the encoder with a model trained without making the encoder dilated. Here follows a diagram:

![make_dilated](https://user-images.githubusercontent.com/17425914/79581136-4a81ef80-80ca-11ea-9966-2314001e79df.png)
In your implementation (2nd cell of the diagram), you also make dilated the first residual units of the block which had a stride of 2, **this is wrong** because it does not preserve the receptive field of the pretrained model (1st cell of the diagram). The solution is presented in the 3rd cell of the diagram.

This occurs because you did not strictly reproduce the code of [torchvision.models.resnet](https://github.com/pytorch/vision/blob/7b60f4db9707d7afdbb87fd4e8ef6906ca014720/torchvision/models/resnet.py#L177) where the dilation rate is set to `previous_dilation` for the first residual unit which had a stride of 2.

<pre>
def _make_layer(self, block, planes, blocks, stride=1, dilate=False):
    norm_layer = self._norm_layer
    downsample = None
    <b>previous_dilation = self.dilation</b>
    if dilate:
        <b>self.dilation *= stride</b>
        stride = 1
    if stride != 1 or self.inplanes != planes * block.expansion:
        downsample = nn.Sequential(
            conv1x1(self.inplanes, planes * block.expansion, stride),
            norm_layer(planes * block.expansion),
        )

    layers = []
    layers.append(block(self.inplanes, planes, stride, downsample, self.groups,
                        self.base_width, <b>previous_dilation</b>, norm_layer))
    self.inplanes = planes * block.expansion
    for _ in range(1, blocks):
        layers.append(block(self.inplanes, planes, groups=self.groups,
                            base_width=self.base_width, <b>dilation=self.dilation</b>,
                            norm_layer=norm_layer))

    return nn.Sequential(*layers)
</pre>

**Comment by qubvel** [comment]
Hi @howard-mahe 

Thanks for clarification, but could you explain what is the motivation behind `preserving receptive field` ? We still training all these conv layers, weights are used just as initialization and features are general, not specific. Do you think this 1 pixel padding will affect a lot for model accuracy? I did not find any inforamtion about `dilation=1` for first block fo `layer` in paper of deeplab. 

On the other hand, such kind of `make_dilated` allow me to use this function to other encoders, without writing custom code for each one.

**Comment by howard-mahe** [comment]
Thank you @qubvel for your interest!

In the paper entitled _"DeepLab: Semantic Image Segmentation with Deep Convolutional Nets, Atrous Convolution, and Fully Connected CRFs"_ which describes `DeepLabV2`, we can read at §3.1:
> For example, in order to double the spatial density of computed feature responses in the VGG-16 or ResNet-101 networks, we find the last pooling or convolutional layer that decreases resolution (’pool5’ or ’conv5 1’ respectively), set its stride to 1 to avoid signal decimation, and replace all
**subsequent** convolutional layers with atrous convolutional layers having rate r = 2.

We need two distinguish two cases:
1. The encoder is fine-tuned i.e. its pretrained weights are just as initialization
2. The encoder is frozen i.e. its weights pretrained on ImageNet are not fine-tuned

In the first case, I agree that it's likely that the weights will adapt and this may not affect the model accuracy for the semantic segmentation task. However in the second case, we can expect non-negligible impact because for the layer that were used to decrease the resolution, the receptive field for a given neuron has doubled i.e. it has changed from `3x3` to `5x5` (as you can see in my diagram). Normally, the purpose of making a dilated backbone is to increase the spatial density without modifying the activations that were already computed for a lower spatial density, But here this purpose is unmet.

I fully understand your wish to keep a `make_dilated` method common to all encoders. I think it is still feasible following the instructions in DeepLab paper and its implementation in some encoders e.g. [slim/mobilenets](https://github.com/tensorflow/models/blob/a174bf5b1db0e2c1e04697ff5aae5182bd1c60e7/research/slim/nets/mobilenet/mobilenet.py#L255).

I will try to propose a patch for the `make_dilated` method tomorrow.

**Comment by qubvel** [comment]
Thank you for the detailed answer, I hope you can make a patch :+1: . 
If you will face any problems or questions, please ask. I can make a perforamnce comparison, I have a pipeline for training models on public datasets and going to release it soon and release pretrained models too (you can see some of them [here](https://github.com/qubvel/segmentation_models.pytorch/tree/feature/dvc/zoo)).

**Comment by howard-mahe** [comment]
Great initiative to use DVC for your zoo, I am not particularly familiar with it but I have no doubt about its usefulness. Thank you for this contribution!

I extensively thought about coding the patch but I no longer think it is feasible to patch the `make_dilated` parent method without making strong assumption or without discriminating multiples cases e.g. plain nets, residual nets, inception nets, etc. Because when we loop over the `layer.modules()`, most times we cannot distinguish sequential paths from parallel paths (e.g. residue with downsampling in resnets), hence it is hard to know where to apply the dilation rate.

For instance, this is `resnet34.layer3`:
```
Sequential(
  (0): BasicBlock(
    (conv1): Conv2d(128, 256, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
    (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace=True)
    (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (downsample): Sequential(
      (0): Conv2d(128, 256, kernel_size=(1, 1), stride=(2, 2), bias=False)
      (1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
  )
  (1): BasicBlock(
    (conv1): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn1): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    (relu): ReLU(inplace=True)
    (conv2): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (bn2): BatchNorm2d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  )
)
```
Here, the first BasicBlock has two parallel path where the strided ops are: `conv1` and `downsample[0]`. We must apply dilation rate on `conv2` but not on `downsample[0]` even if `downsample[0]` looks subsequent to `conv2`, because it is not.

If you are interested at some point by fixing this issue, my recommendation is to follow other library by:
1. supporting the `make_dilated` feature during model build (in model's `__init__` called by `get_encoder`)
2. implementing a `make_dilated` in child encoder classes

**Comment by howard-mahe** [comment]
Hello @qubvel, I was thinking again about this issue and before implementing a patch, I think it would make sense to perform a benchmark as you proposed.

**Comment by andrewsonga** [comment]
@howard-mahe Thank you for your detailed explanation on the correct way to implement `make_dilated`. I was just wondering, if the `stride` for the first 3x3 convolution of a BasicBlock is originally 1, then how should the `stride` and `dilation` be set in the 3rd cell of the diagram? 

I'm asking because I don't quite understand the logic of setting the dilation rate to `previous_dilation` for the first residual unit, which has a stride of 2, in `torchvision.models.resnet`.

Thank you in advance :)

**Comment by howard-mahe** [comment]
@andrewsonga Thank you for your interest in this issue.

First, let me remind that the default value is `dilation=1` which means no dilation actually. As long as you did not encounter a `Conv 3x3 s=2` that you want to make dilated, one must maintain `previous_dilation=1` (default value). To answer to your first question, if the `stride` for the first `Conv 3x3` is 1, one maintain the value of `previous_dilation` and if `previous_dilation` was `1`, then there is no change in the `BasicBlock`.

The logic of setting the dilation rate to `previous_dilation` for **any** residual unit with `Conv 3x3 s=2` is defined by LC Chen himself:  
> we find the last pooling or convolutional layer that decreases resolution (’pool5’ or ’conv5 1’ respectively), set its stride to 1 to avoid signal decimation, and replace all **subsequent** convolutional layers with atrous convolutional layers having rate r = 2.

If you don't conform to this logic, you won't preserve the receptive field of the pretrained model for all neurons. You can have a look to the very first figure of this discussion where the receptive field is depicted in blue. In this figure, you can't see how the receptive is modified if you don't conform to the logic.

**Comment by andrewsonga** [comment]
@howard-mahe Thank you for your detailed response! I now have a better understanding of this design logic.

The reason I asked this question was because I'm trying to apply the Multi-grid method introduced in [1], [2] to ResNet34, whose layers are comprised of the `BasicBlock`, which contains two 3x3 conv layers. My goal is to apply a `Multi_Grid = (r1, r2, r3) = (1, 2, 4)` with a base rate of 2 to the final stage of Resnet34 _i.e.,_ `self.layer4`, _while_ maintaining an output stride (OS) = 16. 

In order to maintain an output stride of OS = 16, we change the stride in the 1st 3x3 conv of the 1st BasicBlock in self.layer4 from 2 to 1. But then in that case, I'm wondering what is the appropriate way of applying the set of dilations 2 * (1, 2, 4) to the 3 `BasicBlocks` of `self.layer4`. Is it ok to set the `i`th dilation to the 2nd 3x3 conv of the `i`th `BasicBlock`?
